cpp/src/barretenberg/stdlib/primitives/field/field.cpp

```diff
@@ -21,9 +21,16 @@ template <typename ComposerContext>
 field_t<ComposerContext>::field_t(const witness_t<ComposerContext>& value)
     : context(value.context)
 {
-    additive_constant = 0;
-    multiplicative_constant = 1;
-    witness_index = value.witness_index;
+    if constexpr (IsSimulator<ComposerContext>) {
+        additive_constant = value.witness;
+        multiplicative_constant = 1;
+        witness_index = IS_CONSTANT;
+        // context -> stored_value = value.witness;
+    } else {
```

Comment:
Normally values of field elements are stored in the `variables` array of a given circuit builder. The index in that array is called a "real  variable index". In order to avoid redundantly storing variable values, a witness object wraps an index `idx` and the corresponding variable value is accessed as in `variables[real_variable_index[idx]]`. I'm trying out a model where we do away with the variables array altogether, but just have a witness wrap a value. Therefore to initialize a field element from a witness I need to supply the witness info not through providing a witness index, but a witness value itself. The right way to do that is through setting the `additive_constant`, since 'normalizing' the field element would put the same value there. That said, I'm not sure this is the best approach.

Reply:
But yeah if you're just talking about the `constexpr` then I think we could avoid code generation by just defining the function for the particular template parameter `CircuitSimulatorBN254`. We will also have a `CircuitSimulatorGrumpkin` though... anyway, experimenting.